### PR TITLE
Hipp 974 autopublish service layer

### PR DESCRIPTION
**Pull request by llmikeyj**

### CODE REVIEW

File: app/uk/gov/hmrc/integrationcatalogueautopublish/services/AutopublishService.scala [hunk 1]
```diff
@@ -18,16 +18,47 @@ package uk.gov.hmrc.integrationcatalogueautopublish.services
...
-  def autopublish(): Future[Either[AutopublishException, Unit]] = {
+  def autopublish()(implicit hc: HeaderCarrier): Future[Either[AutopublishException, Unit]] = {
     logger.info("Starting auto-publish")
-    Future.successful(Right(()))
+    oasDiscoveryConnector.allDeployments() flatMap {
+      case Right(deployments) =>
+        Future.sequence(deployments.map(deployment => {
+          apiRepository.findByPublisherReference(deployment.id) flatMap {
+            case Some(api) if api.deploymentTimestamp.isBefore(deployment.deploymentTimestamp) =>
+              publishAndUpsertRepository(api)
```
Comment (PaulCDurham) [inline app/uk/gov/hmrc/integrationcatalogueautopublish/services/AutopublishService.scala:41]: Need to update the timestamp

File: app/uk/gov/hmrc/integrationcatalogueautopublish/services/AutopublishService.scala [hunk 2]
```diff
@@ -18,16 +18,47 @@ package uk.gov.hmrc.integrationcatalogueautopublish.services
...
+            case None =>
+              publishAndUpsertRepository(Api(deployment.id, deployment.deploymentTimestamp))
+            case _ => Future.successful(Right(()))
+          }
+        })).flatMap(_ => Future.successful(Right(())))
+
+      case Left(_) =>
+        Future.successful(Right())
```
Comment (PaulCDurham) [inline app/uk/gov/hmrc/integrationcatalogueautopublish/services/AutopublishService.scala:49]: We should probably return an exception as we can't do anything at all here

File: app/uk/gov/hmrc/integrationcatalogueautopublish/services/AutopublishService.scala [hunk 3]
```diff
@@ -18,16 +18,47 @@ package uk.gov.hmrc.integrationcatalogueautopublish.services
...
+    }
+  }
+
+  private def publishAndUpsertRepository(api: Api)(implicit hc: HeaderCarrier): Future[Either[AutopublishException, Unit]] = {
+    oasDiscoveryConnector.oas(api.publisherReference) flatMap {
+      case Right(oasDocument) => integrationCatalogueConnector.publishApi(oasDocument.id, oasDocument.content) flatMap {
+        case Right(()) =>
+          apiRepository.upsert(api).flatMap(_ => Future.successful(Right(())))
```
Comment (PaulCDurham) [inline app/uk/gov/hmrc/integrationcatalogueautopublish/services/AutopublishService.scala:57]: Maybe log a warning if the repo update fails?

**Review by PaulCDurham — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/integrationcatalogueautopublish/services/AutopublishService.scala (hunk 1 above)
Comment (llmikeyj) [reply app/uk/gov/hmrc/integrationcatalogueautopublish/services/AutopublishService.scala]: Ok

File: app/uk/gov/hmrc/integrationcatalogueautopublish/services/AutopublishService.scala (hunk 2 above)
Comment (llmikeyj) [reply app/uk/gov/hmrc/integrationcatalogueautopublish/services/AutopublishService.scala]: ok

**Pull request merged**